src/webgpu/util/check_contents.ts

```diff
@@ -0,0 +1,234 @@
+      failedElements[i] = true;
+    }
+  }
+
+  if (failedElementsFirstMaybe === undefined) {
+    return undefined;
+  }
+  const failedElementsFirst = failedElementsFirstMaybe;
+  const failedElementsLast = failedElements.length - 1;
+
+  // Include one extra non-failed element at the beginning and end (if they exist), for context.
+  const printElementsStart = Math.max(0, failedElementsFirst - 1);
+  const printElementsEnd = Math.min(size, failedElementsLast + 2);
+  const printElementsCount = printElementsEnd - printElementsStart;
+
+  const numberToString = printAsFloat
+    ? (n: number) => n.toPrecision(4)
+    : (n: number) => intToPaddedHex(n, { byteLength: ctor.BYTES_PER_ELEMENT });
+  const numberPrefix = printAsFloat ? '   ' : '0x:';
+
+  const printActual = actual.subarray(printElementsStart, printElementsEnd);
+  const printExpected: Array<Iterable<string | number>> = [];
+  if (predicatePrinter) {
+    for (const { leftHeader, cell } of predicatePrinter) {
+      printExpected.push(
+        (function* () {
+          yield* [leftHeader, numberPrefix];
+          yield* iterRange(printElementsCount, i => cell(printElementsStart + i));
+        })()
+      );
+    }
+  }
+
+  const printFailedValueMarkers = (function* () {
+    yield* ['failed ->', ''];
+    yield* range(printElementsCount, i => (failedElements[printElementsStart + i] ? 'xx' : ''));
+  })();
+
+  const opts = {
+    fillToWidth: 120,
+    numberToString,
+  };
+  const msg = `Array had unexpected contents at indices ${failedElementsFirst} through ${failedElementsLast}.
+ Starting at index ${printElementsStart}:
+${generatePrettyTable(
+  opts,
+  ['actual ==', numberPrefix, ...printActual],
+  printFailedValueMarkers,
+  ...printExpected
+)}`;
+  return new ErrorWithExtra(msg, () => ({
+    actual: actual.slice(),
+  }));
+}
+
+// Helper helpers
+
+/** Convert an integral `number` into a hex string, padded to the specified `byteLength`. */
+function intToPaddedHex(number: number, { byteLength }: { byteLength: number }) {
+  assert(Number.isInteger(number), 'number must be integer');
+  let s = Math.abs(number).toString(16);
+  if (byteLength) s = s.padStart(byteLength * 2, '0');
+  if (number < 0) s = '-' + s;
+  return s;
+}
+
+/**
+ * Pretty-prints a "table" of cell values (each being `number | string`), right-aligned.
+ * Each row may be any iterator, including lazily-generated (potentially infinite) rows.
+ *
+ * The first argument is the printing options:
+ *  - fillToWidth: Keep printing columns (as long as there is data) until this width is passed.
+ *    If there is more data, "..." is appended.
+ *  - numberToString: if a cell value is a number, this is used to stringify it.
+ *
+ * Each remaining argument provides one row for the table.
+ */
+function generatePrettyTable(
+  { fillToWidth, numberToString }: { fillToWidth: number; numberToString: (n: number) => string },
+  ...rows: ReadonlyArray<Iterable<string | number>>
```

Comment:
nit: we could pass a list directly here instead of relying on `...` this seems quite obscure in JS/TS.